viewer/image/Image.cpp

```diff
@@ -41,12 +41,69 @@ float Image::value(int x, int y, int channel) const
   }
 }
 
+Result<Image> Image::scaleByHalf() const
+{
+  // Looses some pixels for odd resolutions
+
+  if (width_ <= 1 && height_ <= 1) {
+    return Result<Image>("Image is too small for further downscaling by half.");
+  }
+
+  int newWidth = std::max(width_ / 2, 1);
+  int newHeight = std::max(height_ / 2, 1);
+
+  std::vector<uint8_t> newdata(newWidth * newHeight * channels_ * pixelSizeInBytes());
+  if (format_ == Float)
+  {
+    const float* i = reinterpret_cast<const float*>(data());
+    float* o = reinterpret_cast<float*>(newdata.data());
```

Comment:
Both these casts violate strict aliasing rule. Although extremely unlikely, it's still UB and the compiler may decide to fuck you up. See suggestions below.